Ability to draw a zone's side faces, not just outline

braem opened the pull request:
Closes #909 

- Drawn very similarly to zone outlines, using `debugoverlay`.
- Only draws side faces as bottom/top faces are complex with `debugoverlay`.
- Only draws outline/faces if the zone if a start zone or on current track.
- Cvars for each zone type a la the outline.

Gross thing is that it draws through walls a bit, and also the duration is arbitrary. 0 duration like the outline causes flickering, so does values larger or smaller than 0.001. Doesn't seem to be framerate dependant. Larger values also work but then lags the game a lot.

### Checklist
- [x] **I have thoroughly tested all of the code I have modified/added/removed to ensure something else did not break**
- [x] If there is a localization token change, I have updated the `momentum_english_ref.res` file with the changes, ran `tokenizer.py` to generate an up-to-date localization file, and have committed both the `.res` file changes and the new localization `.txt` file
- [x] If I introduced new h/cpp files, I have added them to the appropriate project's VPC file (`server_momentum.vpc` / `client_momentum.vpc` / etc)
- [x] If I have added or modified any visual assets (models, materials, panels, effects, etc), I have taken screenshots / videos of them and attached them to this PR directly (screenshots uploaded through github, videos uploaded to youtube and linked)
- [x] If I have modified any console command, console variable, or momentum entity, I have opened an issue (or a PR) for it in the [Momentum Mod documentation repository](https://github.com/momentum-mod/docs)
- [x] My commits are relatively small and scoped to the best of my ability
- [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
- [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review

Gocnak commented:
Duplicating the info here so it's not lost in the sauce, but the CMeshBuilder option should work so long as you fake it having transparency:

From SCell:
> the entity itself has to lie about being transparent, and/or maybe 2-pass
override IsTransparent(IsTwoPass) calls

braem commented:
Do you happen to have the PR for the time where you moved the implementation to `debugoverlay`? @Gocnak 
Mainly looking for the code to create the models needed to move everything off `debugoverlay`.

Also feel free to tackle this yourself especially if you'd like this in 0.8.6. I'm pretty outside my realm here so it will be a bit.

braem commented:
> 
> 
> Duplicating the info here so it's not lost in the sauce, but the CMeshBuilder option should work so long as you fake it having transparency:
> 
> From SCell:
> 
> > the entity itself has to lie about being transparent, and/or maybe 2-pass
> > override IsTransparent(IsTwoPass) calls

This resolved the issue you were seeing with the outline not rendering.

braem commented:
- Completely got off `debugoverlay` for rendering zone outlines, thanks to SCell's fix.
- Added a method for drawing side faces using mesh builder. Problem here though as alphas lower than 180 just cause it to not render at all for some reason.
- Moved away from toggles to an enum; one for outline, one for faces using mesh builder, and one for faces using `debugoverlay`. Did this as you mentioned it might be useful to have this overlay version so you can see zones through walls.
  - When there is a model, it draws the entire polygon if it's either of the mesh builder or `debugoverlay` face enums.
- Added ability to override zone face alphas as I believe people will want transparent faces but solid outlines.

Most of the weird things like not drawing full polygons when there isn't a model, alphas lower than 180 causing mesh builder not to work, etc. are due to my lack of knowledge with it not intentional design.

Also there is a lot of Z fighting with mesh builder despite it rendering as two pass & transparent.. I don't know if that's normal or what.

SCell555 commented on mp/src/game/client/momentum/c_mom_triggers.cpp:
```diff
@@ -83,61 +110,163 @@ C_BaseMomZoneTrigger::C_BaseMomZoneTrigger()
-        Vector curUp(cur.x, cur.y, cur.z + m_flZoneHeight);
-        Vector nextUp(next.x, next.y, next.z + m_flZoneHeight);
-        debugoverlay->AddLineOverlayAlpha(curUp, nextUp, outlineColor.r(), outlineColor.g(), outlineColor.b(), outlineColor.a(), false, 0.0f);
+        const Vector next(m_vecZonePoints[i].x, m_vecZonePoints[i].y, m_vecZonePoints[i].z + m_flZoneHeight);
```
You do not need to construct Vector to pass it to Position3fv, just use Position3f and pass each component directly

SCell555 replied:
No. You call it with `builder.Position3f(x, y, z)`, in force push it was changed correctly in DrawSideFacesModelAsBrush, but not here

SCell555 commented on mp/src/game/client/momentum/c_mom_triggers.cpp:
```diff
@@ -83,61 +110,161 @@ C_BaseMomZoneTrigger::C_BaseMomZoneTrigger()
+
+    for (int i = 0; i < iNum; i++)
+    {
+        const auto vecCurr = m_vecZonePoints[i];
```
These also should be refs
```suggestion
        const auto& vecCurr = m_vecZonePoints[i];
```

SCell555 commented on mp/src/game/client/momentum/c_mom_triggers.cpp:
```diff
@@ -83,61 +110,161 @@ C_BaseMomZoneTrigger::C_BaseMomZoneTrigger()
+
+    for (int i = 0; i < iNum; i++)
+    {
+        const auto vecCurr = m_vecZonePoints[i];
+        const auto vecNext = m_vecZonePoints[(i + 1) % iNum];
```
These also should be refs
```suggestion
        const auto& vecCurr = m_vecZonePoints[i];
        const auto& vecNext = m_vecZonePoints[(i + 1) % iNum];
```

braem commented on mp/src/game/client/momentum/c_mom_triggers.cpp:
```diff
@@ -49,18 +75,19 @@ bool CTriggerOutlineRenderer::RenderBrushModelSurface(IClientEntity* pBaseEntity
     builder.Begin(pRenderContext->GetDynamicMesh(true, nullptr, nullptr, 
-                                                 materials->FindMaterial("momentum/zone_outline", TEXTURE_GROUP_OTHER)),
-                    MATERIAL_LINE_LOOP, vertices);
+                  materials->FindMaterial("momentum/zone_outline", TEXTURE_GROUP_OTHER)),
```
Precached in `CMOMClientEvents::LevelInitPreEntity` in `client_events`. `CMOMClientEvents::Precache()` is never called for some reason.

braem commented on mp/src/game/client/momentum/c_mom_triggers.cpp:
```diff
@@ -83,61 +110,161 @@ C_BaseMomZoneTrigger::C_BaseMomZoneTrigger()
+                case MOM_ZONE_DRAW_MODE_FACES_OVERLAY:
+                    DrawSideFacesModelAsOverlay(m_ZoneModelRenderer.m_Color);
+                    break;
                 }
```
Yeah they dont, I removed the parameter.

braem commented:
Did the requested changes & made sure everything still works fine (didn't lose something in rebase).

Gocnak commented on mp/src/game/client/momentum/c_mom_triggers.cpp:
```diff
@@ -9,18 +9,44 @@
+static MAKE_CONVAR(mom_zone_checkpoint_draw_mode, "1", FCVAR_CLIENTCMD_CAN_EXECUTE | FCVAR_ARCHIVE,
+                   "Changes the drawing mode for checkpoint zones.\n 0 = Off; None, 1 = Outline (Default), 2 = Side faces as brushes, 3 = Side faces as overlays", MOM_ZONE_DRAW_MODE_FIRST, MOM_ZONE_DRAW_MODE_LAST);
+
+static ConVar mom_zone_start_draw_color("mom_zone_start_draw_color", "00FF00FF", FCVAR_CLIENTCMD_CAN_EXECUTE | FCVAR_ARCHIVE,
+                                        "Color of the start zones.");
```
Can be one line and you forgot your newlines at the end of all of these:
```suggestion
static ConVar mom_zone_start_draw_color("mom_zone_start_draw_color", "00FF00FF", FCVAR_CLIENTCMD_CAN_EXECUTE | FCVAR_ARCHIVE, "Color of the start zones in RGBA hex.\n");
```

Gocnak commented on mp/src/game/client/momentum/client_events.cpp:
```diff
@@ -107,6 +107,7 @@ void CMOMClientEvents::PostInit()
 void CMOMClientEvents::LevelInitPreEntity()
 {
     //Precache();
+    PrecacheMaterial("momentum/zone_outline");
```
Does this effectively remove the `Binding uncached material "momentum/zone_outline"` warning in console if you go on triggertests and turn around at spawn?

Also, we may want to keep this in the triggers class so that the material name can be pulled out to a macro.

If it's not precaching, you'll have to make sure `Precache()` being called inside the base trigger's `Spawn()` function.

braem replied:
Yes it does remove the warning which is nice. I'll move it to the client triggers and pull it into a macro.

braem replied:
Doing it this way removes the warning, but not on the first load. The `Precache()` happens before calling `DrawModel` yet it still has this error when rendering for the first time. I'm confused as to why.

For now I'll just push my changes as it's really simple to switch back if we want.

Gocnak replied:
It probably needs to be precached on the server first. I ran into this with the conc branch recently, a material would not have a model index until it was done on server.

braem replied:
Wondering if its better to be in `mom_triggers` as well or just where it was before

Gocnak replied:
Can be in mom_triggers yes.

Gocnak commented on mp/src/game/client/momentum/c_mom_triggers.cpp:
```diff
@@ -80,64 +109,167 @@ C_BaseMomZoneTrigger::C_BaseMomZoneTrigger()
+    if (!pRunEntData)
+        return 0;
+
+    const auto pStartZone = dynamic_cast<C_TriggerTimerStart*>(this);
+    if (!pStartZone && m_iTrackNumber != pRunEntData->m_iCurrentTrack)
```
Not a really big fan of doing a dynamic cast every render call. A better solution here would be to network [the zone type](https://github.com/momentum-mod/game/blob/develop/mp/src/game/server/momentum/mom_triggers.cpp#L166) across to here and check that instead. You can set the zone type in the constructor of each zone trigger.

Gocnak replied:
Or, instead of networking this, you could just do the exact same code (virtual base class that all the children override and return the right value). Might be better that way, yeah.

braem replied:
Did the latter

Gocnak commented on mp/src/game/client/momentum/c_mom_triggers.cpp:
```diff
@@ -80,64 +109,167 @@ C_BaseMomZoneTrigger::C_BaseMomZoneTrigger()
+    if (iRenderMode && flags & STUDIO_RENDER && (flags & (STUDIO_SHADOWDEPTHTEXTURE | STUDIO_SHADOWDEPTHTEXTURE)) == 0)
     {
-        if ((flags & STUDIO_RENDER) && (flags & (STUDIO_SHADOWDEPTHTEXTURE | STUDIO_SHADOWDEPTHTEXTURE)) == 0)
+        if (GetDrawColor())
```
Can pull this out to the outer if statement above it.

Gocnak reviewed (dismissed):
Can you get some photos of the new rendering in the initial PR message?

Also, does this fully close the linked issue?

braem commented:
> 
> 
> Can you get some photos of the new rendering in the initial PR message?
> 
> Also, does this fully close the linked issue?

Outline rendering (`mom_zone_*_draw_mode 1`), outlines always rendered like this:
https://cdn.discordapp.com/attachments/722973307292614687/768003045589385226/SPOILER_2020-10-19_23-44-41.mp4
Did on surf_mesa to show that rendering issue you mentioned doesn't happen.

Side face rendering with `debugoverlay` (`mom_zone_*_draw_mode 3` with `.zon` file):
https://cdn.discordapp.com/attachments/722973307292614687/768002930955255808/2020-10-19_23-47-46.mp4
Best side face rendering in my opinion. Video also shows the logic of only showing zones on current track & all start zones.

Side face rendering with `CMeshBuilder` (`mom_zone_*_draw_mode 2` with `.zon` file):
https://cdn.discordapp.com/attachments/722973307292614687/768003508262666250/2020-10-19_23-50-56.mp4
- disappears with an alpha value lower than 180
- Z-fighting (not shown in vid)

Polygon rendering with `CMeshBuilder` (`mom_zone_*_draw_mode 2` & `mom_zone_*_draw_mode 3` with in-map zones):
https://cdn.discordapp.com/attachments/722973307292614687/768003854775353344/2020-10-19_23-52-35.mp4
Draws as a polygon. Has the same setbacks as the other `CMeshBuilder` option. Video shows the bad Z-fighting.

This does not fully close the issue given the aforementioned bugs that I don't know how to solve, particularly with `CMeshBuilder` rendering. Also I don't know how to draw the tops & bottoms of the zones, just the side faces. Polygon drawing works easily with in-map zones as the vertices are in order to do so.

I changed from "Closes" to "Closes part of"

SCell555 commented:
You could try editing zone_outline.vmt to match what debugoverlay system uses
```
UnlitGeneric
{
	$vertexcolor 1
	$vertexalpha 1
}
```

braem commented:
> 
> 
> You could try editing zone_outline.vmt to match what debugoverlay system uses
> 
> ```
> UnlitGeneric
> {
> 	$vertexcolor 1
> 	$vertexalpha 1
> }
> ```

The material uses this though there are more fields here. I am able to get rid of Z fighting by setting `$nocull` to `0`, and stop it from disappearing on lower alpha values with `$alphatest` as `0`.

Also has an `$ignorez` field that changes whether it can be seen through walls. I can completely remove `debugoverlay` with this by using a separate material that has a different `$ignorez` value.

SCell555 commented:
You might be able to draw 2 quads, one normal and the other with mirrored vertex order. It still should be cheaper than `debugoverlay->AddBoxOverlay`, as that draws 3d box with one side 0 thick

braem commented:
Pushed a commit that adds a zone outline as overlay drawing option and also moves side faces as overlay rendering off of `debugoverlay`. Done using 2 materials: one for regular rendering and another for overlay-style (see through walls).

Kept the old commits in case we need to go back to them.

Did precaching in `mom_triggers` and put the zone material's paths in shareddefs. Not sure if this is the best way.

> 
> 
> You might be able to draw 2 quads, one normal and the other with mirrored vertex order. It still should be cheaper than `debugoverlay->AddBoxOverlay`, as that draws 3d box with one side 0 thick

Is this referencing drawing the tops and bottoms? I couldn't get it to work because the ordering of the vertices doesn't nicely allow for it.

SCell555 commented:
Can you try if this will work? [test.diff.txt](https://github.com/momentum-mod/game/files/5411275/test.diff.txt)

braem commented:
> 
> 
> Can you try if this will work? [test.diff.txt](https://github.com/momentum-mod/game/files/5411275/test.diff.txt)

Works for a lot of zones but not all it seems:
![image](https://user-images.githubusercontent.com/9014762/96644614-2b188900-12de-11eb-813a-c9a40c73bdcb.png)

SCell555 commented:
Does this also happen to side faces? Or is it only top/bottom

braem commented:
> 
> 
> Does this also happen to side faces? Or is it only top/bottom. Because this should not happen, looks like there are some extra points in zone file. Unless this happens only with face rendering, not outlines.

This was the issue I was experiencing trying to do this. Outlines render fine:
![image](https://user-images.githubusercontent.com/9014762/96651128-ffe76700-12e8-11eb-9f4e-13b5448fe6f2.png)

Issue is only with drawing bottom & top faces; side faces appear to be fine.

SCell555 commented:
I think that I have a solution for this problem, but it won't work if zone shape is concave

braem commented:
Looks good with the overlay material
![image](https://user-images.githubusercontent.com/9014762/96658857-7d67a300-12fa-11eb-9d8b-ec28c32aa778.png)

but not with the other:
![image](https://user-images.githubusercontent.com/9014762/96659229-4219a400-12fb-11eb-8824-8f47721fef83.png)

Seems its showing overlapping textures where `$ignorez` doesnt. Maybe something I can change in the vmt?

SCell555 commented:
I forgot to swap lines 207 and 305 when I was copying that code block around

braem commented:
> 
> 
> I forgot to swap lines 207 and 305 when I was copying that code block around

Eyy would you look at that! Brilliant :D 

Drawing concave stuff only looks a little weird too, nothing super major:
![image](https://user-images.githubusercontent.com/9014762/96744145-9499a580-1379-11eb-9217-fa33852804f2.png)

Feel free to commit this to this branch if you want, I would feel bad taking any sort of credit commit-wise.

Gocnak commented:
You should be able to co-author people through GitHub desktop at least.

braem commented:
> 
> 
> You should be able to co-author people through GitHub desktop at least.

Pushed it with that. Was trying to avoid my name altogether but I suppose you can't do that unless SCell committed directly.
